cpp/include/Ice/SlicedData.h

```diff
@@ -59,6 +60,14 @@ namespace Ice
               hasOptionalMembers(hasOptionalMembers),
               isLastSlice(isLastSlice)
         {
+            if (compactId == -1)
+            {
+                assert(typeId.empty());
+            }
+            else
+            {
+                assert(!typeId.empty());
+            }
```

Comment:
```suggestion
            assert((compactId == -1) == typeId.empty());
```

Single assert and the compiler no if/else blocks for the compiler to feel tempted to leave in.